### PR TITLE
stm32wb55/powerctrl: Add BLE support to deepsleep.

### PR DESCRIPTION
**Pull request by andrewleech**
We've found a rather thorny WB55 issue that I've been investigating for much of the past week - deepsleep (standby).

After deepsleep is used, BLE cannot be enabled again. 

A colleague first hit this on one of our active projects trying to use low power support. 
In his case (on mpy 1.18) after this failure occurs even a reset will not fix it, neither machine.reset() nor physical reset switch, it needs a hard power off / on.

```
MicroPython v1.18-129-g1133d1ee5-dirty on 2022-02-21; NUCLEO-WB55 with STM32WB55RGV6
Type "help()" for more information.
>>> import machine, bluetooth
>>> ble = bluetooth.BLE()
>>> ble.active(1)
>>> ble.gap_advertise(20_000, b'\x02\x01\x06\x04\tMPY')
>>> ble.active(0)
>>> machine.deepsleep(1000)
MicroPython v1.18-129-g1133d1ee5-dirty on 2022-02-21; NUCLEO-WB55 with STM32WB55RGV6
Type "help()" for more information.
>>> import machine, bluetooth;ble = bluetooth.BLE();ble.active(1);ble.gap_advertise(20_000, b'\x02\x01\x06\x04\tMPY');ble.active(0)
rfcore: rfcore_ble_init
tl_sys_wait_ack: timeout
tl_ble_wait_resp: timeout
tl_sys_wait_ack: timeout
rfcore: rfcore_ble_hci_cmd
rfcore: rfcore_ble_init
tl_sys_wait_ack: timeout
tl_ble_wait_resp: timeout
tl_sys_wait_ack: timeout
rfcore: rfcore_ble_hci_cmd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 110] ETIMEDOUT
```

I've tested on ble hci radio firmware 1.12 and 1.13.1 and both appear to behave the same, though this current commit here has only been tested on 1.13.1 so far. 

There's supposed to be extra code needed to wake Core2 after standby as the PWR_CR4_C2BOOT flag only works on cold boot.
https://community.st.com/s/question/0D53W00000A28aPSAR/stm32wb55-implement-standby-mode-in-a-ble-application
Just adding this interrupt trigger isn't enough to solve the issue for me though.

I've got a number of changes and commented out tests in this PR, largely inspired by posts on the stm forums and this official example project which has sleep / suspend support: https://github.com/STMicroelectronics/STM32CubeWB/tree/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_HeartRate 

As it stands in the current commit, deepsleep support appears to be fixed, as long as BLE is disabled. I've got this set up to automatically disable BLE whenever deepsleep is called.

Alternatively, BLE can be left enabled (with future potential to have it used to wake from sleep) with boot code to detect the post-standby startup, at which point NVIC_SystemReset() is run. This does "solve" the timeout problem, but is feels like a pretty nasty hack (even if it's a suggested workaround by STM employees on forum posts).

On a side note, if I replace HAL_PWR_EnterSTANDBYMode() with HAL_PWR_EnterSHUTDOWNMode() the problem is also fixed, and you get slightly lowever power use - however BLE cannot be left enabled in this case, so the fix as committed here is likely better.

On a related note, this solution seems quite fragile. Pretty much all the individual changes included here appear to be required, swapping out pretty much any of them brings the problem back. I suspect there could be timing issues too, if the BLE deinit / shutdown isn't done early enough before standby is entered, the problem remains.

I haven't had any luck finding a means to inspect core2 / ble firmware to know what state it's in? If it's possible to check / know when it's disabled / low power before interring standby on core1 that would be better?

Separately, it'd be awesome to add a multi-test for this feature - however I've failed to get one to work so far as the test harness drops out when the chip reboots. Has any work been done previously anywhere to add support to auto re-connect and continue a test? Or should we extend mpremote soft-reboot handling to also detect sleep/reboots and then use it for muiltitest?

### CODE REVIEW

**Comment by andrewleech** [comment]
Ok, I've got this cleaned up to a much simpler, easier to understand changeset.

While I'm still not certain why usb connected / not connected was causing a difference in behaviour, I suspect it's possibly just a startup timing isse - would like to look into this more.

Either way, I found that capturing the initial timeout and using that to trigger a re-run of rfcore_init() fixes the issue in all situations; it recovers from the timeout as well as resets the ipcc queue to remove the hard fault. It does mean that in the post-deepsleep startup we get one of the timeout messages printed to console, not sure if we want to make that a quiet / debug printout?

Also, somewhat strangely, I need to enable HCI on startup or I still get the timeout in a way that rfcode init does not recover from. Disabling it again after HSE is enabled seems fine. Again, this is only needed on post-standby boot, not on cold boot.

**Comment by andrewleech** [comment]
Even better, I found a fix for the initial timeout with a little extra semaphore synchronisation. So now the `if (timeout) rfcore_init()` probably isn't needed as much, however it might be a good fallback to have in case the semaphore doesn't always work.

**Comment by andrewleech** [comment]
Notes for reference:
* https://www.st.com/resource/en/application_note/an5289-building-wireless-applications-with-stm32wb-series-microcontrollers-stmicroelectronics.pdf
* https://github.com/STMicroelectronics/STM32CubeWB/blob/0b858907b3cf836c051b674c38e69f01968d4e9e/Projects/NUCLEO-WB15CC/Applications/BLE/BLE_DataThroughput/Core/Src/standby_stm32wb15.c#L106
* https://github.com/STMicroelectronics/STM32CubeWB/blob/master/Projects/STM32WB5MM-DK/Applications/BLE/BLE_TransparentMode/Core/Src/stm32_lpm_if.c

**Comment by andrewleech** [comment]
Ok, I've brought in some clearer RCC semaphore management as per AN5289, following the reference design in https://github.com/STMicroelectronics/STM32CubeWB/blob/master/Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_DataThroughput/Core/Src/stm32_lpm_if.c

Configuring these semaphores & clocks before entering sleep has removed the need to toggle HSI on/off during startup which is fantastic as that was one of the main changes I'd made previously that never felt right.

I've also cleaned up the rest of the code with proper `#ifdefs` etc so should be ready for final review.

File: ports/stm32/boards/stm32wbxx_hal_conf_base.h [hunk 1]
```diff
@@ -64,19 +65,56 @@
...
 #define assert_param(expr) ((void)0)
 
+// Hardware Semaphores - ref: AN5289
+
+// Index of the semaphore used the prevent conflicts after standby sleep.
+// Each CPUs takes this semaphore at standby wakeup until conclicting elements are restored.
+// Note: this is used in WB15 reference examples, but not listed in AN5289 Rev 6
+#define CFG_HW_PWR_STANDBY_SEMID 10
+
+/* Ensures that CPU2 does not update the BLE persistent data in SRAM2 when CPU1 is reading them */
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32wbxx_hal_conf_base.h:90]: please use `//` comments

File: ports/stm32/powerctrlboot.c [hunk 2]
```diff
@@ -281,6 +280,9 @@ void SystemClock_Config(void) {
 
     SystemCoreClockUpdate();
     powerctrl_config_systick();
+
+    /* Release RCC semaphore */
```
Comment (dpgeorge) [inline ports/stm32/powerctrlboot.c:284]: please use `//`

File: tests/multi_bluetooth/ble_lightsleep.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# Test BLE GAP connect/disconnect
```
Comment (dpgeorge) [inline tests/multi_bluetooth/ble_lightsleep.py:1]: please rename test to `ble_deepsleep.py` if it's for deepsleep

Comment (andrewleech) [reply tests/multi_bluetooth/ble_lightsleep.py:1]: and I should add some comments about the limitations when running this test (needs uart repl etc)

Comment (andrewleech) [reply tests/multi_bluetooth/ble_lightsleep.py]: Done now. I remembered this test was unfinished - because it didn't work with the instance0 processor rebooting from the deep sleep. I've now added support for thie in the mult-test runner, let me know what you think.

File: tests/run-multitests.py [hunk 4]
```diff
@@ -90,6 +90,15 @@ def get_network_ip():
             except:
                 ip = HOST_IP
         return ip
+    @staticmethod
+    class will_reboot:
```
Comment (dpgeorge) [inline tests/run-multitests.py:94]: I think it's simpler to make this a function `multitest.expect_reboot(...)`.

File: tests/run-multitests.py [hunk 5]
```diff
@@ -378,6 +407,15 @@ def run_test_on_instances(test_file, num_instances, instances):
                 last_read_time[idx] = time.time()
                 if out is not None and not any(m in out for m in IGNORE_OUTPUT_MATCHES):
                     trace_instance_output(idx, out)
+                    if out.startswith("WAIT_FOR_REBOOT"):
+                        _, resume, delay_ms = out.split(" ")
```
Comment (dpgeorge) [inline tests/run-multitests.py:411]: the resume string should be a full name, eg `instance0_resume`.  Then verify that the prefix `instance0` matches what's expected.

File: tests/run-multitests.py [hunk 6]
```diff
@@ -406,6 +444,54 @@ def run_test_on_instances(test_file, num_instances, instances):
...
+    BANNER_TIMEOUT = 2
+    QUIET_TIMEOUT = 0.1
+    FULL_TIMEOUT = 5 + extra_timeout
+    t_start = t_last_activity = time.monotonic()
+    data_all = ""
+    reboot_banner = False
+    next_data_timeout = QUIET_TIMEOUT
+    while True:
+        t = time.monotonic()
+        out = instance.read(1)
```
Comment (dpgeorge) [inline tests/run-multitests.py:460]: Try to simplify this to use `instance.readline()` only.

File: tests/multi_bluetooth/ble_deepsleep.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    with multitest.will_reboot(resume="0_resume", delay_ms=100):
+        machine.deepsleep(100)
+
+
+# Acting in peripheral role.
+def instance0_resume():
+    # continue here after reboot
+    print("restarted")
+    global ble, irq
+    multitest.globals(BDADDR=ble.config("mac"))
```
Comment (dpgeorge) [inline tests/multi_bluetooth/ble_deepsleep.py:63]: this line should be removed:
- it's not needed because it's already set in `instance0()`, and it'll be the same value
- `instance1()` will not pick up the changed value (if it were to change) because it's still running the code with the old value injected into the templace script
- it appears in the expected output when it should not (because the BDADDR changes with a different MCU)

**Comment by andrewleech** [comment]
Thanks for the pointers @dpgeorge I've addressed those changes in the multitests.

Now, on firmware without this change:
```
anl@STEP: ~/micropython/tests $ python3 run-multitests.py -i pyb:/dev/ttyACM1 -i micropython multi_bluetooth/ble_deepsleep.py -t
multi_bluetooth/ble_deepsleep.py on ttyACM1|micropython:
TRACE ttyACM1|micropython:
   573 i0 : SET BDADDR = (0, b"\x02\x02'NB\xce")
   573 i0 : gap_advertise
   574 i0 : NEXT
   676 i1 : NEXT
   677 i0 : _IRQ_CENTRAL_CONNECT
   677 i1 : gap_connect
   677 i1 : _IRQ_PERIPHERAL_CONNECT
   677 i1 : gap_disconnect: True
   779 i0 : _IRQ_CENTRAL_DISCONNECT
   779 i1 : _IRQ_PERIPHERAL_DISCONNECT
   779 i0 : WAIT_FOR_REBOOT instance0_resume 100
  3356 i1 : gap_connect
  3661 i0 : tl_sys_wait_ack: timeout
  3967 i0 : tl_ble_wait_resp: timeout
  4171 i0 : tl_sys_wait_ack: timeout
  4780 i1 : Traceback (most recent call last):
  4780 i1 :   File "<stdin>", line 155, in <module>
  4780 i1 :   File "<stdin>", line 91, in instance1
  4780 i1 :   File "<stdin>", line 40, in wait_for_event
  4781 i1 : ValueError: Timeout waiting for 7
```
The multitest actually hangs at this point because i0 continues to spam `tl_sys_wait_ack: timeout` lines forever.

With this PR applied:
```
anl@STEP: ~/micropython/tests $ python3 run-multitests.py -i pyb:/dev/ttyACM1 -i micropython multi_bluetooth/ble_deepsleep.py -t
multi_bluetooth/ble_deepsleep.py on ttyACM1|micropython:
TRACE ttyACM1|micropython:
   571 i0 : SET BDADDR = (0, b"\x02\x02'NB\xce")
   571 i0 : gap_advertise
   571 i0 : NEXT
   674 i1 : NEXT
   674 i0 : _IRQ_CENTRAL_CONNECT
   674 i1 : gap_connect
   674 i1 : _IRQ_PERIPHERAL_CONNECT
   674 i1 : gap_disconnect: True
   776 i0 : _IRQ_CENTRAL_DISCONNECT
   776 i1 : _IRQ_PERIPHERAL_DISCONNECT
   776 i0 : WAIT_FOR_REBOOT instance0_resume 100
  1445 i1 : gap_connect
  1546 i0 : restarted
  1547 i1 : _IRQ_PERIPHERAL_CONNECT
  1547 i0 : gap_advertise
  1548 i0 : _IRQ_CENTRAL_CONNECT
  1549 i0 : gap_disconnect: True
  1854 i0 : _IRQ_CENTRAL_DISCONNECT
  4388 i1 : _IRQ_PERIPHERAL_DISCONNECT
pass
1 tests performed
1 tests passed
```

This was tested with a NUCLEO_WB55 on i0 and unix port with nrf52810 dongle as usbserial hci ble as i1.

**Comment by dpgeorge** [comment]
Thanks, looks really good!

Testing locally with NUCLEO_WB55 and PYBD_SF2, it's working well.

**Comment by dpgeorge** [comment]
This doesn't work when mboot is used.  I pushed a fix for that.

File: ports/stm32/rfcore.c [hunk 8]
```diff
@@ -589,18 +600,34 @@ static const struct {
...
     DEBUG_printf("rfcore_ble_init\n");
 
-    // Clear any outstanding messages from ipcc_init.
-    tl_check_msg(&ipcc_mem_sys_queue, IPCC_CH_SYS, NULL);
-
     // Configure and reset the BLE controller.
-    tl_sys_hci_cmd_resp(HCI_OPCODE(OGF_VENDOR, OCF_BLE_INIT), (const uint8_t *)&ble_init_params, sizeof(ble_init_params), 0);
-    tl_ble_hci_cmd_resp(HCI_OPCODE(0x03, 0x0003), NULL, 0);
+    if (!rfcore_ble_reset()) {
+        // Out of standby initial init can fail. Reset rfcode and retry.
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:605]: I don't think this re-init-on-fail is needed anymore.  I tested without it and it seems OK.

But do you want to keep it in anyway, just in case?

File: ports/stm32/powerctrlboot.c [hunk 9]
```diff
@@ -28,6 +28,29 @@
...
+        RCC->CIER = 0x00000000;
+    } else {
+        // Other start-up (eg POR), use standard system init code.
+        SystemInit();
+    }
+}
+#else
+void stm32_system_init(void) {
+    // TODO find a way to eliminate this simple function.
+    SystemInit();
```
Comment (dpgeorge) [inline ports/stm32/powerctrlboot.c:50]: I don't like this wasted function that just calls another one (in critical start-up code that needs to be fast).

The only way I can think to eliminate it (on non WB55) is to have a new `resethandler.s` that's used on MCUs that want to intercept the `SystemInit()` function.

File: ports/stm32/rfcore.c (hunk 8 above)
Comment (andrewleech) [reply ports/stm32/rfcore.c:605]: Yeah I was flipping back and forth on keeping it or not... I think it was really just to fix a specific problem that's already solved better now elsewhere. 
I say remove this retry; If there's ever a different problem that this _could_ workaround it's probably better we know it's happening and solve it properly too.

File: ports/stm32/powerctrlboot.c (hunk 9 above)
Comment (andrewleech) [reply ports/stm32/powerctrlboot.c:50]: Couldn't you just 
```
#define  stm32_system_init SystemInit
```

Comment (dpgeorge) [reply ports/stm32/powerctrlboot.c:50]: No, because `stm32_system_init` is called from assembler, `resethandler.s`.

File: ports/stm32/rfcore.c [hunk 10]
```diff
@@ -589,18 +600,38 @@ static const struct {
...
     // Configure and reset the BLE controller.
-    tl_sys_hci_cmd_resp(HCI_OPCODE(OGF_VENDOR, OCF_BLE_INIT), (const uint8_t *)&ble_init_params, sizeof(ble_init_params), 0);
-    tl_ble_hci_cmd_resp(HCI_OPCODE(0x03, 0x0003), NULL, 0);
+    if (!rfcore_ble_reset()) {
+        // ble init can fail if core2 has previously locked up. Reset HSI & rfcore to retry.
+        __HAL_RCC_HSI_DISABLE();
+        mp_hal_delay_ms(100);
+        __HAL_RCC_HSI_ENABLE();
+
+        rfcore_init();
```
Comment (andrewleech) [inline ports/stm32/rfcore.c:610]: I found a reason to keep an extended retry in - another team at work has had intermittent issues with this same "timeout" symptom on new WB55 chips after they're initially programmed. Only happens on some units, not on most. I suspect there's a similar timing / race-condition of HSI clock on/off/on during programming then restart, first boot.

I've extended this retry to also include a HSI off, wait, on again. 

I got my board into the "permanently locked up state" where it would show timeout on `BLE().active(1)` even after hard reset. Then reflashed with this change and  `BLE().active(1)` now works, after showing a single instance of the timeout message.

File: ports/stm32/rfcore.c (hunk 8 above)
Comment (andrewleech) [reply ports/stm32/rfcore.c]: https://github.com/micropython/micropython/pull/8337#discussion_r1127333834

File: ports/stm32/powerctrlboot.c (hunk 9 above)
Comment (andrewleech) [reply ports/stm32/powerctrlboot.c]: I've refactored this to call `stm32_system_init` / `SystemInit` from the start of `stm32_main()` instead, removing the empty wrapper function.

On first try it failed for me... deepsleep still put it into timeout state. Then I realised I needed to re-compile and flash mboot too! Once that was done it all works perfectly for me thanks!

File: ports/stm32/main.c [hunk 11]
```diff
@@ -426,6 +428,9 @@ void stm32_main(uint32_t reset_mode) {
     #endif
     systick_enable_dispatch(SYSTICK_DISPATCH_LWIP, mod_network_lwip_poll_wrapper);
     #endif
+    #if defined(STM32WB)
+    rfcore_init();
```
Comment (dpgeorge) [inline ports/stm32/main.c:432]: Why was this moved down here?  It should be fine here, just curious.

File: ports/stm32/boards/stm32wbxx_hal_conf_base.h [hunk 12]
```diff
@@ -64,19 +65,56 @@
...
-#define DATA_CACHE_ENABLE           1
-#define INSTRUCTION_CACHE_ENABLE    1
-#define PREFETCH_ENABLE             0
-#define USE_SPI_CRC                 0
-#define USE_RTOS                    0
+#define DATA_CACHE_ENABLE 1
+#define INSTRUCTION_CACHE_ENABLE 1
+#define PREFETCH_ENABLE 0
+#define USE_SPI_CRC 0
+#define USE_RTOS 0
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32wbxx_hal_conf_base.h:78]: best to leave these lines untouched

File: ports/stm32/boards/stm32wbxx_hal_conf_base.h [hunk 13]
```diff
@@ -64,19 +65,56 @@
...
 
 // HAL parameter assertions are disabled
 #define assert_param(expr) ((void)0)
 
+// Hardware Semaphores - ref: AN5289
+
+// Index of the semaphore used the prevent conflicts after standby sleep.
+// Each CPUs takes this semaphore at standby wakeup until conclicting elements are restored.
+// Note: this is used in WB15 reference examples, but not listed in AN5289 Rev 6
+#define CFG_HW_PWR_STANDBY_SEMID 10
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32wbxx_hal_conf_base.h:88]: maybe indent tho numbers to the right to line up?

File: ports/stm32/main.c (hunk 11 above)
Comment (andrewleech) [reply ports/stm32/main.c:432]: Ah, leftover from an earlier trial when I thought the startup timing of this might be a key impact... yeah I don't have a reason to keep it moved.

File: ports/stm32/boards/stm32wbxx_hal_conf_base.h (hunk 12 above)
Comment (andrewleech) [reply ports/stm32/boards/stm32wbxx_hal_conf_base.h]: I think it was codeformat.py that "cleaned" that up, I've fixed these and the new ones as suggested, see what CI here things of it :-)

File: ports/stm32/rfcore.c [hunk 14]
```diff
@@ -589,18 +600,38 @@ static const struct {
...
-    tl_check_msg(&ipcc_mem_sys_queue, IPCC_CH_SYS, NULL);
-
     // Configure and reset the BLE controller.
-    tl_sys_hci_cmd_resp(HCI_OPCODE(OGF_VENDOR, OCF_BLE_INIT), (const uint8_t *)&ble_init_params, sizeof(ble_init_params), 0);
-    tl_ble_hci_cmd_resp(HCI_OPCODE(0x03, 0x0003), NULL, 0);
+    if (!rfcore_ble_reset()) {
+        // ble init can fail if core2 has previously locked up. Reset HSI & rfcore to retry.
+        __HAL_RCC_HSI_DISABLE();
+        mp_hal_delay_ms(100);
+        __HAL_RCC_HSI_ENABLE();
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:608]: maybe use `LL_RCC_HSI_Enable()` etc, to be consistent with the new code added in `powerctrl.c`

File: ports/stm32/powerctrl.c [hunk 15]
```diff
@@ -762,6 +796,29 @@ void powerctrl_enter_stop_mode(void) {
...
+        __HAL_FLASH_SET_LATENCY(FLASH_LATENCY_1);
+        while (!LL_RCC_HSE_IsReady()) {
+        }
+        LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSE);
+        while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_STATUS_HSE) {
+        }
+    }
+
+    // Release RCC semaphore
+    LL_HSEM_ReleaseLock(HSEM, CFG_HW_RCC_SEMID, 0);
```
Comment (dpgeorge) [inline ports/stm32/powerctrl.c:819]: maybe factor this code to a function like the other code, eg `powerctrl_low_power_exit_wb55()`?

File: ports/stm32/rfcore.c (hunk 14 above)
Comment (andrewleech) [reply ports/stm32/rfcore.c:608]: Yeah... sorry I did think I should check the implementation of both of them to confirm if the HAL variant does more useful stuff over the LL one to select which is actually most appropriate here. These HAL ones are used elsewhere already too...

File: ports/stm32/boards/stm32wbxx_hal_conf_base.h [hunk 16]
```diff
@@ -42,6 +42,7 @@
 #include "stm32wbxx_hal_uart.h"
 #include "stm32wbxx_hal_usart.h"
 #include "stm32wbxx_ll_adc.h"
+#include <stm32wbxx_ll_hsem.h>
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32wbxx_hal_conf_base.h:45]: please use `"` to match existing includes

File: ports/stm32/main.c (hunk 11 above)
Comment (dpgeorge) [reply ports/stm32/main.c]: OK, then let's keep it where it was originally

File: ports/stm32/rfcore.c (hunk 10 above)
Comment (dpgeorge) [reply ports/stm32/rfcore.c:610]: OK, let's keep it

File: ports/stm32/rfcore.c (hunk 14 above)
Comment (dpgeorge) [reply ports/stm32/rfcore.c:608]: The LL ones are fine to use.  I think they are exactly what's needed and nothing more.

Comment (andrewleech) [reply ports/stm32/rfcore.c:608]: Oh hah, well the stm impl for the hal functions doesn't add much...
```
#define __HAL_RCC_HSI_ENABLE()  LL_RCC_HSI_Enable()
#define __HAL_RCC_HSI_DISABLE() LL_RCC_HSI_Disable()
```

**Comment by dpgeorge** [comment]
I modified a little my commit that provides a custom `SystemInit`, so that the selection of wb55 / no-wb55 happens in one location in powerctrl.h, rather than in both main.c and mboot/main.c.

**Comment by dpgeorge** [comment]
Merged!  Thanks @andrewleech  for all your efforts on this.

**Comment by andrewleech** [comment]
Thanks for the additional improvements, support and testing @dpgeorge !
